[bench] enable mlir benchmark

xurui1995 opened the pull request:
Enable benchmark and tuning from python
trace: https://github.com/intel/graph-compiler/issues/57

- [x] running gc-cpu-pipeline in python
- [x] bench a MLIR file with path
- [x] generate and bench mlp pattern with options
- [x] implement tuner  (will have a new PR)
- [x] support tensor or memref with different data types 

There are still some features that will be implemented when integrated into the PR of benchgc https://github.com/intel/graph-compiler/pull/161 ，such as data filling, customize pass pipeline.

Please read this if you need to use this benchmark and tuner tools.
https://github.com/intel/graph-compiler/blob/xurui/add_benchmark/tools/README.md

xurui1995 commented:
# Content of README

# Python Tools
## Pre-requisites
### Enable python binding
* Enable MLIR python binding, [README](https://github.com/intel/graph-compiler/blob/main/python/README.md)
* Install `/tools/requirements.txt`
### Set env
* **PYTHONPATH**=*${BUILD_DIR}*/python_packages/gc_mlir_core
* **LD_PRELOAD**=path/to/libiomp5.so
* **MLIR_C_RUNNER_UTILS**=*${LLVM_INSTALL_DIR}*/lib/libmlir_c_runner_utils.so
* **MLIR_RUNNER_UTILS**=*${LLVM_INSTALL_DIR}*/lib/libmlir_runner_utils.so


## Bench
The tool has two different ways to calculate the time cost, and more experiments are needed to test which one is more stable and accurate. Currently, users can choose which way to use through options
* Use the MLIR Python API to invoke the kernel and use Python to calculate the time cost
* Modify MLIR by wrapping the kernel into a new method and calling the `nanoTime()` method before and after calling the kernel. Finally, calculate the difference as the time cost
```
  func.func private @nanoTime() -> i64 attributes {llvm.emit_c_interface}
  func.func public @wrapped_main(%arg0: memref<1xi64>, %arg1: tensor<128x512xbf16>, %arg2: tensor<512x256xbf16>) -> tensor<128x256xbf16> attributes {llvm.emit_c_interface} {
    %0 = call @nanoTime() : () -> i64
    %1 = call @main_entry(%arg1, %arg2) : (tensor<128x512xbf16>, tensor<512x256xbf16>) -> tensor<128x256xbf16>
    %2 = call @nanoTime() : () -> i64
    %3 = arith.subi %2, %0 : i64
    %c0 = arith.constant 0 : index
    memref.store %3, %arg0[%c0] : memref<1xi64>
    return %1 : tensor<128x256xbf16>
  }
}
```

### Examples:
```
# simple version
python3 ./tools/main.py --driver=load_mlir --path=./tools/workloads/test.mlir

# complex version
python3 ./tools/main.py --type=bench --bench_alg=py --driver=load_mlir --path=./tools/workloads/test.mlir --warm_up=200 --repeat=200 --print_ir --entry=main_entry
```

```
# result example
===========bench result===========
{
    "args": {
        "type": "bench",
        "driver": "load_mlir",
        "path": "./tools/workloads/test.mlir",
        "entry": "main_entry",
        "bench_alg": "py",
        "print_ir": false,
        "warm_up": 20,
        "repeat": 100
    },
    "compile_cost(ms)": 25.58841183781624,
    "execute_cost(ms)": 1.7501823976635933
}
```

### Common Options
*  `--driver`: the pattern to bench, currently support `mlp` and `load_mlir`
*  `--bench_alg`: `py` or `wrapper`, different evaluation implementation of the benchmark
*  `--warm_up`: warm-up times of the execution
*  `--repeat`: repeat times of the execution
*  `--print_ir`: print the ir before execution
*  `--disable_results_to_params`: do not use this when using the default pipeline (gc-cpu-pipeline)

### Driver Specific Options
* load_mlir
  * `--path`: the mlir file path
  * `--entry`: the name of entry func
```
python3 ./tools/main.py --driver=load_mlir --path=./tools/workloads/test.mlir
```


* mlp  
  * `--batch_size`: the input
  * `--hidden_size_list`: hidden_sizes of mlp, example: 32x16x64
  * `--has_bias`: if the matmul op has bias, example: 1x0
  * `--act_type`: choices=["noop", "relu", "sigmoid"]
  * `--dtype`: choices=["bf16", "f32"]
```
python3 ./tools/main.py --driver=mlp --batch_size=32 --hidden_size_list=32x16x64 --has_bias=0x0 --act_type=noop --dtype=f32

===========bench func name:  main_entry ===========
module {
  func.func @main_entry(%arg0: tensor<32x32xf32>, %arg1: tensor<32x16xf32>, %arg2: tensor<16x64xf32>) -> tensor<32x64xf32> attributes {llvm.emit_c_interface} {
    %0 = tensor.empty() : tensor<32x16xf32>
    %1 = linalg.matmul {cast = #linalg.type_fn<cast_signed>} ins(%arg0, %arg1 : tensor<32x32xf32>, tensor<32x16xf32>) outs(%0 : tensor<32x16xf32>) -> tensor<32x16xf32>
    %2 = tensor.empty() : tensor<32x64xf32>
    %3 = linalg.matmul {cast = #linalg.type_fn<cast_signed>} ins(%1, %arg2 : tensor<32x16xf32>, tensor<16x64xf32>) outs(%2 : tensor<32x64xf32>) -> tensor<32x64xf32>
    return %3 : tensor<32x64xf32>
  }
}
```

## Tuning
The logic of tuner is consistent with that of graph compiler v1 version, which can generate different config based on user set candidates and constraints
### Examples
```
python3 /home/xurui/gc_v2/graph-compiler/tools/main.py  --driver=load_mlir --type=tune --path=./tools/workloads/test.mlir --bench_alg=wrapper  --search_alg=grid --batch_size=50
```

```
# result
[ 50 / 125 ] skipped: 0 best: 0.8382204799999994 ms
[ 100 / 125 ] skipped: 0 best: 0.83801604 ms
[ 125 / 125 ] skipped: 0 best: 0.83801604 ms
Tuning ends in 28.885201930999756 s
Best cost: 0.83801604 ms
Best config: [{
    "MatMulConfig": {
        "M_threads": 1,
        "K_threads": 1,
        "N_threads": 1,
        "M_block": 128,
        "K_block": 128,
        "N_block": 32,
        "innermostM_block": 1,
        "innermostK_block": 1,
        "innermostN_block": 1
    }
}]
mlir:
 module {
  func.func @main_entry(%arg0: tensor<128x512xbf16>, %arg1: tensor<512x256xbf16>) -> tensor<128x256xbf16> attributes {llvm.emit_c_interface} {
    %cst = arith.constant 0.000000e+00 : bf16
    %0 = tensor.empty() : tensor<128x256xbf16>
    %1 = linalg.fill ins(%cst : bf16) outs(%0 : tensor<128x256xbf16>) -> tensor<128x256xbf16>
    %2 = linalg.matmul {KBlock = 128 : i32, Kthreads = 1 : i32, MBlock = 128 : i32, Mthreads = 1 : i32, NBlock = 32 : i32, Nthreads = 1 : i32, innermostKBlock = 1 : i32, innermostMBlock = 1 : i32, innermostNBlock = 1 : i32} ins(%arg0, %arg1 : tensor<128x512xbf16>, tensor<512x256xbf16>) outs(%1 : tensor<128x256xbf16>) -> tensor<128x256xbf16>
    return %2 : tensor<128x256xbf16>
  }
}
```

### Options
The tuner will share the option of the bench tool and has the following unique tuner options
* `--tuning_batch`: The tuner will execute after generating a batch number of config, default: 50
* `--timeout`: The time limit setting for the tuner, default: -1
* `--max_tuning_iters`: The maximum number of iterations for tuning can also be considered as the maximum number of configs executed, default: sys.maxsize
* `--space_percent`：The percentage of tuning space is used, default: 1.0
* `--checkpoint_path`: The path to save the tuner status so that you can continue the tuning next time
* `--search_alg`: The algorithm for searching configs, choices=["grid", "ga"]
* `--early_stop`: If the tuner does not find a better config after iterating the early stop value, it will stop, default: -1, never stop

- When using the `ga`
  * `--random_seed`: Seed value for random function in GA tuner
  * `--elite_num`: the concept of elite number in genetic algorithm, default: 9
  * `--mutation_prob`：concept of mutation rate in genetic algorithm, default: 0.1
  * `--expected_tune_num`: The approximate quantity of generated config, If the value is 0, the ga tuner will use HashSet to record the tuned configs, otherwise use bloom filter, default: 0

Menooker commented:
Are there any overlap with #161? Which do you suggest to review first? @WangJialei-A  @xurui1995

ciyongch commented on tools/bench.py:
```diff
@@ -0,0 +1,220 @@
+
+    func = engine.lookup(entry_name)
+    packed_args = (ctypes.c_void_p * len(mlir_args))()
+    for argNum in range(len(mlir_args)):
```
`argNum` -> `arg_num` or `arg_idx`?

xurui1995 replied:
I copied the code from MLIR[ execution_engine.py](https://github.com/llvm/llvm-project/blob/4283f1ad18db9878b98f98e7a36b4f94ab674d29/mlir/python/mlir/execution_engine.py#L26-L35) so that the cast cost does not affect performance result, and it's more accurate than call the `execution_engine.invoke` directly. Even the naming is not good, I prefer to keep it, and i will add a comment here.

ciyongch commented on tools/main.py:
```diff
@@ -0,0 +1,209 @@
+    )
+    add_driver_args(parser)
+    parser.add_argument(
+        "--bench_alg", type=str, choices=["py", "wrapper"], default="py"
```
`bench_kind`?

xurui1995 replied:
fixed

ciyongch commented on tools/tuner.py:
```diff
@@ -0,0 +1,603 @@
+            return
+        start_time = time.time()
+        spaces_size = self.tunning_space.space_size
+        while self.iter < max_iter and self.iter < spaces_size:
```
Do we have plan to support parallelized tuning?

xurui1995 replied:
yes, It's on the to-do list, will add that after integrated into benchgc

xurui1995 commented on test/gc/python/test_compiler.py:
```diff
@@ -0,0 +1,36 @@
+        )
+        compiler.compile(module)
+        #  CHECK-NOT: onednn_graph.matmul
+        print(module)
```
> No need to print the module in the test, but adding some other check with expected result instead?

The ` print(module)` result was used for the FileCheck, for example here the output was used for `#  CHECK-NOT: onednn_graph.matmul`

xurui1995 replied:
> I don't get why we would run lits through python bindings. Why can't we cover such tests with regular lits?

The porpuse here is to test if the gc dialects and passes are bound correctly, similar to the test dir `llvm-project/mlir/test/python/*.py` in MLIR.

Menooker replied:
Shall we just use python's own unittest modules to do binding tests, if the cases are really about python itself, instead of the text of MLIR.
Please note that `lit` and `FileCheck` are two independent tools from LLVM. I agree that we use `lit` to drive all the tests for both `*.mlir` and `*.py`. However, we don't necessarily need to use FileCheck in this case. `lit` just call this script and check the exit code of the script.
You are free to use FileCheck or python's own unittest for python checkings. You can choose the most fit approach for your case.

xurui1995 replied:
Sorry, I did not notice the folder of this `test_compiler.py`, and now it was removed, and the test was added into `smoketest.py`.
For the FileCheck part, the python tests under `graph-compiler/test/mlir/test/gc/python/`  using it as the way in the MLIR `llvm-project/mlir/test/python/`, if we have dialects to upstream in the future, the python API tests are also necessary, so we would better using the same check.
Lastly, for the unittest modules  to do binding tests, I got your point, but I think for now it's fine as a part of `gc-check`, just like the python test in MLIR is also a part of `chek-mlir`,

WangJialei-A commented on tools/README.md:
```diff
@@ -0,0 +1,151 @@
+### Set env
+* **PYTHONPATH**=*${BUILD_DIR}*/python_packages/gc_mlir_core
+* **LD_PRELOAD**=path/to/libiomp5.so
+* **MLIR_C_RUNNER_UTILS**=*${LLVM_INSTALL_DIR}*/lib/libmlir_c_runner_utils.so
+* **MLIR_RUNNER_UTILS**=*${LLVM_INSTALL_DIR}*/lib/libmlir_runner_utils.so
```
@xurui1995 
We may need an elegant way to handle this dependency?
Copying the related shared object to the `build` directory during the `cmake` step is another option?

xurui1995 replied:
@kurapov-peter @WangJialei-A  thanks for the advice, I will handle this in the next PR for migrating the code here into about the benchgc folder  https://github.com/intel/graph-compiler/issues/172,

xurui1995 replied:
fixed now, get the path from `MLIR_DIR` by cmake

kurapov-peter commented on tools/bench.py:
```diff
@@ -0,0 +1,220 @@
+    )
+    total_time = 0
+    ns_to_ms_scale = 1e-6
+    def run(engine_invoke, bench_func_name, *mlir_args):
+        engine_invoke(bench_func_name, *mlir_args)
+
+    for i in range(repeat_time + warm_up):
+        run(engine.invoke, "wrapped_main", time_arg, *mlir_args)
+        if i >= warm_up:
+            total_time += int(np_timers_ns[0]) * ns_to_ms_scale
+    execute_cost = total_time / repeat_time
```
Just a note, we'll need cold run measurements as well

kurapov-peter commented on tools/bench.py:
```diff
@@ -0,0 +1,220 @@
+        pipeline,
+        shared_libs,
+    )
+    compile_begin = timeit.default_timer()
+    engine = compiler.compile_and_jit(wrapper_module, ir_printing=ir_printing)
+    compile_cost = (timeit.default_timer() - compile_begin) * 1000
```
Can be a decorator

xurui1995 replied:
do you mean add a decorator for the `compile_and_jit function`? if so, is there a way to get both `engine` and `compile_cost ` ?

xurui1995 replied:
sry, I'm still confused about the decorator, is the follwing decorator something you expect?

```
def timing_decorator(func):
    def wrapper(*args, **kwargs):
        start_time = timeit.default_timer()
        result = func(*args, **kwargs)
        end_time = timeit.default_timer()  
        elapsed_time = (end_time - start_time) * 1000
        return elapsed_time, result
    return wrapper


# add decorator for compile_and_jit
@timing_decorator
def compile_and_jit(
    elf, module: ir.Module, ir_printing: bool = False
) -> execution_engine.ExecutionEngine:
    self.compile(module, ir_printing)
    return self.jit(module)


# invoke
compile_cost, engine = compiler.compile_and_jit(ir_module, ir_printing=ir_printing)

```

kurapov-peter commented on tools/config_filter.py:
```diff
@@ -0,0 +1,84 @@
+        pass
+
+
+class BloomFilter(ConfigFilter):
```
What are these for?

Menooker replied:
They are for the tuner. I guess we don't really need this currently. `ConfigFilter` will check if a config (generated from tuner) is already met before and skip it. `BloomFilter` uses O(1) space to check the existence of "unlimited“ number elements. We can remove this for now..

kurapov-peter commented on tools/tuner.py:
```diff
@@ -0,0 +1,603 @@
+
+class TuningSpace:
+    """
+    The class works as a bride between the tuner and the configs in MLIR module.
```
:))
```suggestion
    The class works as a bridge between the tuner and the configs in MLIR module.
```

xurui1995 commented on tools/README.md:
```diff
@@ -0,0 +1,151 @@
+
+### Common Options
+*  `--driver`: the pattern to bench, currently support `mlp` and `load_mlir`
+*  `--bench_alg`: `py` or `wrapper`, different evaluation implementation of the benchmark
```
will change to `bench_kind`.
do you mean provide both results for A single benchmark run？

xurui1995 commented on python/gc_mlir/dialects/linalgxOps.td:
it's how MLIR does the python bindings, for each dialect, they have td and py files in the python folder for the bindinds. 
for linalg: https://github.com/llvm/llvm-project/blob/main/mlir/python/mlir/dialects/LinalgOps.td

xurui1995 commented on lib/gc/Transforms/OneDNNGraphToLinalg.cpp:
```diff
@@ -492,6 +492,8 @@ struct MatMulOpLowering : public OpRewritePattern<MatMulOp> {
     }
 
+    // Passing mutmal configs to linalg.matmul
+    newOp->setAttrs(op->getAttrs());
```
removed now

Menooker commented:
> The tuner can go in separately

I agree with you. tuner is an additional feature built on benchmark. We can limit the scope of this PR into the benchmark, to make it easier  and faster to review, and to unblock the other PR that depends on benchmark.

xurui1995 commented:
> The tuner can go in separately

I will remove the tuner code in this PR

xurui1995 commented:
@kurapov-peter  hi, peter, could you help review the changes?

kurapov-peter reviewed (commented):
I still find the purpose of the tool puzzling. Why does it rely on FileCheck to run IR tests? What does it add on top of just running FileCheck via lits?

xurui1995 commented:
> I still find the purpose of the tool puzzling. Why does it rely on FileCheck to run IR tests? What does it add on top of just running FileCheck via lits?

The FileCheck in python is for testing if the bindings work, this is also how MLIR tests their bindings， and it's not the purpose of this bench tool.

xurui1995 commented on test/mlir/test/gc/python/dialects/linalgx.py:
To test if the Linalgx binding works, If we don't do the binding properly, the `Module.parse` will get an error.

ZhennanQin commented on tools/bench.py:
We did this just because the benchmarking function is easy to implement, just like the implementation here: https://github.com/intel/graph-compiler-playground/blob/main/dl_bench/benchmark.py#L106-L141. We are not against using any existing tools if they can meet our requirements. Do you have any recommendation?

xurui1995 replied:
Firstly it's not just timing python calls here, as mentioned before, there are two types of bench kinds supported here:
* `bench kind=py`: Use the MLIR Python API to invoke the kernel and use Python to calculate the time cost
* `bench kind=wrapper`: Modify MLIR by wrapping the kernel into a new method and calling the nanoTime() method before and after calling the kernel. Finally, calculate the difference as the time cost

As for why not use existing bench tools, the reason is I found methodology of them is simple, they use the similar way of `bench kind=wrapper`,  and implementing it ourselves is more convenient for expansion.

ZhennanQin commented:
> > I still find the purpose of the tool puzzling. Why does it rely on FileCheck to run IR tests? What does it add on top of just running FileCheck via lits?
> 
> The FileCheck in python is for testing if the bindings work, this is also how MLIR tests their bindings， and it's not the purpose of this bench tool.

Seems you put two features together:

1. Add python-binding support for GC downstream dialect and passes.
2. Enable benchmark tool on top of python-binding.

Would you please split them into separate PRs for better code-review?

xurui1995 commented on tools/requirements.txt:
```diff
@@ -0,0 +1 @@
+ml_dtypes    # for bf16 in numpy
```
Previously, I had a PR in the MLIR community about this; it was approved and merged, but due to some post-CI issues, it was reverted. if they fixed their CI issues and re-merge the PR， we will not need to add the dependency here. 

In addition to this, we are also trying to use torch tensors to replace numpy  ndarray to avoid this dependency, which will be updated in Jialei's PR https://github.com/intel/graph-compiler/pull/161.

kurapov-peter replied:
> In addition to this, we are also trying to use torch tensors to replace numpy ndarray to avoid this dependency, which will be updated in Jialei's PR #161.

Let's do that first then

xurui1995 replied:
hi, @kurapov-peter ,  the dependency has now been merged into the upstream, as seen here: https://github.com/llvm/llvm-project/pull/101271. Additionally, the recent update to our LLVM version includes that commit, and I have removed the dependency in this PR https://github.com/intel/graph-compiler/pull/106/commits/61fc0d6e2331bb233eb2ba342ee67c53e6a0dd58.

xurui1995 commented:
> > > I still find the purpose of the tool puzzling. Why does it rely on FileCheck to run IR tests? What does it add on top of just running FileCheck via lits?
> > 
> > 
> > The FileCheck in python is for testing if the bindings work, this is also how MLIR tests their bindings， and it's not the purpose of this bench tool.
> 
> Seems you put two features together:
> 
> 1. Add python-binding support for GC downstream dialect and passes.
> 2. Enable benchmark tool on top of python-binding.
> 
> Would you please split them into separate PRs for better code-review?

 linalgx bindings code was removed now

xurui1995 commented on python/CMakeLists.txt:
```diff
@@ -96,6 +108,11 @@ add_mlir_python_modules(GcPythonModules
+  )
+
+set(MLIR_C_RUNNER_UTILS "${MLIR_DIR}/../../libmlir_c_runner_utils.so")
+set(MLIR_RUNNER_UTILS "${MLIR_DIR}/../../libmlir_runner_utils.so")
```
fixed now

xurui1995 commented:
@kurapov-peter @Menooker @ZhennanQin @ciyongch  hi, could you help to review the changes since last time?  @WangJialei-A 's validation PR https://github.com/intel/graph-compiler/pull/161 is blocked now, let's get this merged first.

ZhennanQin commented on python/config.py.in:
```diff
@@ -0,0 +1,2 @@
+MLIR_C_RUNNER_UTILS = "@MLIR_DIR@/../../libmlir_c_runner_utils.so"
```
This is not a good solution for Windows platform. Please try platform platform-independent solution.

Menooker replied:
I agree with @ZhennanQin . Actually there is some examples in our repo for python binding `.py.in` to generate file names with OS-specific extension names. Also `@MLIR_DIR@/../../` seems fragile. I think we have already exposed this directory somewhere in `.py.in` ?

xurui1995 replied:
my laptop has a problem now, I will fix this as soon as it returns to normal. I am trying to change it with `os.path` to implement platform-independent.
@Menooker we do have that exposed but it was in the test dir for lit config, I can not use that from here.  
I also noticed that we have something like the following in other `.py.in`, should change it to the solution with `os.path` ?
```
lit_config.load_config(config, "@PROJECT_SOURCE_DIR@/test/mlir/test/lit.cfg.py")
...
config.substitutions.append(("%gclibdir", config.gc_obj_root + "/lib/"))
...
```

ZhennanQin commented on python/config.py.in:
```diff
@@ -0,0 +1,8 @@
+import os
+
+MLIR_C_RUNNER_UTILS = os.path.normpath(
```
It would be better to just register a folder path here, maybe `MLIR_SHARED_LIB_PATH`, and then in the `graph_compiler.py file`, you can search the lib name in an OS independent way.

Menooker replied:
We can set mlir_runner_utils_dir to `@MLIR_RUNNER_UTILS_DIR@` for our `lit.site.cfg.py.in`. Can we use that cmake variable too?

xurui1995 replied:
I tried and we can not get the `@MLIR_RUNNER_UTILS_DIR@`  from here, not in the same scope.

Menooker commented on python/config.py.in:
```diff
@@ -2,13 +2,12 @@ import os
 
 if sys.platform.startswith("win32"):
-    mlir_runner_utils_dir = os.path.normpath(os.path.join(llvm_obj_root, "bin"))
+    mlir_runner_utils_dir = os.path.normpath(os.path.join("@LLVM_BINARY_DIR@", "bin"))
```
why are we "inlining" this variable? I think using a `llvm_obj_root` to hold the dir name would be better?

xurui1995 replied:
fixed

Menooker commented on python/CMakeLists.txt:
```diff
@@ -112,6 +115,9 @@ add_mlir_python_modules(GcPythonModules
\ No newline at end of file
+  )
+
+configure_file(config.py.in ${MLIR_BINARY_DIR}/python_packages/gc_mlir_core/gc_mlir/config.py @ONLY)
```
you can test to print `message(STATUS "${MLIR_RUNNER_UTILS_DIR}")` to check if `MLIR_RUNNER_UTILS_DIR` is correctly defined here.

xurui1995 replied:
> you can test to print `message(STATUS "${MLIR_RUNNER_UTILS_DIR}")` to check if `MLIR_RUNNER_UTILS_DIR` is correctly defined here.

I tried before and the value of `${MLIR_RUNNER_UTILS_DIR}` is empty here, it was set in `graph-compiler/test/mlir/test/CMakeLists.txt`

Menooker replied:
OK, I see. No further questions here.
